File: solutions/python/determine-if-the-number-is-valid.py
class STATE:
    START, INTEGER, DECIMAL, UNKNOWN, AFTER_DECIMAL = range(5)

def get_next_state(current_state, ch):
  if (current_state == STATE.START or
        current_state == STATE.INTEGER):
    if ch == '.':
      return STATE.DECIMAL
    elif ch >= '0' and ch <= '9':
      return STATE.INTEGER
    else:
      return STATE.UNKNOWN
      
  if current_state == STATE.DECIMAL:
    if ch >= '0' and ch <= '9':
      return STATE.AFTER_DECIMAL
    else:
      return STATE.UNKNOWN
    
  if current_state == STATE.AFTER_DECIMAL:
    if ch >= '0' and ch <= '9':
      return STATE.AFTER_DECIMAL
    else:
      return STATE.UNKNOWN
  return STATE.UNKNOWN

def is_number_valid(s):
  if not s: 
    return True

  i = 0
  if s[i] == '+' or s[i] == '-':
    i = i + 1

  current_state = STATE.START

  for c in s[i:]:
    current_state = get_next_state(current_state, c)
    if current_state == STATE.UNKNOWN:
      return False
    i = i + 1
   
  if current_state == STATE.DECIMAL:
    return False;
  
  return True


def test(s, expected):
  is_valid = is_number_valid(s)
  print(s, is_valid)
  assert is_valid == expected

def main():
  test("4.325", True)
  test("4.325a", False)
  test("x4.325", False)
  test("4.32.5", False)
  test("4325", True)
  test("1.", False)
  test("1.1.", False)
  test("1.1.1", False)
  test("1.1.1.", False)
  test("+1.1.", False)
  test("+1.1", True)
  test("-1.1.", False)
  test("-1.1", True)
  test("", True)
main()

File: solutions/javascript/determine-if-the-number-is-valid.js
const STATE = {
    INTEGER: 'INTEGER',
    DECIMAL: 'DECIMAL',
    START: 'START',
    UNKNOWN: 'UNKNOWN',
    AFTER_DECIMAL: 'AFTER_DECIMAL'
};

let get_next_state = function(current_state, ch) {
    if (current_state === STATE.START ||
        current_state === STATE.INTEGER) {
        if (ch === '.') {
            return STATE.DECIMAL;
        } else if (ch >= '0' && ch <= '9') {
            return STATE.INTEGER;
        } else {
            return STATE.UNKNOWN;
        }
    }
    if (current_state === STATE.DECIMAL) {
        if (ch >= '0' && ch <= '9') {
            return STATE.AFTER_DECIMAL;
        } else {
            return STATE.UNKNOWN;
        }
    }
    if (current_state === STATE.AFTER_DECIMAL) {
        if (ch >= '0' && ch <= '9') {
            return STATE.AFTER_DECIMAL;
        } else {
            return STATE.UNKNOWN;
        }
    }
    return STATE.UNKNOWN;
};

let is_number_valid = function(s) {
    if (s.length === 0) {
        return true;
    }

    let i = 0;
    if (s[i] === '+' || s[i] === '-') {
        i++;
    }


    let current_state = STATE.START;

    for (let l = i; l < s.length; l++) {
        current_state = get_next_state(current_state, s[l]);
        if (current_state === STATE.UNKNOWN) {
            return false;
        }
    }
  
    if (current_state === STATE.DECIMAL)
      return false;
  
    return true;
};

console.log("");
console.log("");
console.log("+++++++++++++++++++++++++++++++++++++++");
console.log("Is Valid Number");
console.log("---------------------------------------");

console.log('is this number valid "4.325" = ' + is_number_valid("4.325"));
console.log('is this number valid "4.325a" = ' + is_number_valid("4.325a"));
console.log('is this number valid "x4.325" = ' + is_number_valid("x4.325"));
console.log('is this number valid "4.32.5" = ' + is_number_valid("4.32.5"));
console.log('is this number valid "4325" = ' + is_number_valid("4325"));
console.log('is this number valid "1." = ' + is_number_valid("1."));
console.log('is this number valid "1.1." = ' + is_number_valid("1.1."));
console.log('is this number valid "1.1.1" = ' + is_number_valid("1.1.1"));
console.log('is this number valid "1.1.1." = ' + is_number_valid("1.1.1."));
console.log('is this number valid "+1.1." = ' + is_number_valid("+1.1."));
console.log('is this number valid "+1.1" = ' + is_number_valid("+1.1"));
console.log('is this number valid "-1.1." = ' + is_number_valid("-1.1."));
console.log('is this number valid "-1.1" = ' + is_number_valid("-1.1"));
console.log('is this number valid "" = ' + is_number_valid(""));
